src/jdk.compiler/share/classes/com/sun/tools/javac/api/JavacTaskImpl.java

```diff
@@ -97,13 +97,18 @@ public Boolean call() {
     /* Internal version of call exposing Main.Result. */
     public Main.Result doCall() {
         try {
-            return handleExceptions(() -> {
+            Pair<Main.Result, Throwable> result = invocationHelper(() -> {
                 prepareCompiler(false);
                 if (compiler.errorCount() > 0)
                     return Main.Result.ERROR;
                 compiler.compile(args.getFileObjects(), args.getClassNames(), processors, addModules);
                 return (compiler.errorCount() > 0) ? Main.Result.ERROR : Main.Result.OK; // FIXME?
-            }, Main.Result.SYSERR, Main.Result.ABNORMAL);
+            });
+            if (result.snd == null) {
+                return result.fst;
+            } else {
+                return (result.snd instanceof FatalError) ? Main.Result.SYSERR : Main.Result.ABNORMAL;
```

Comment:
It would be nicer if we didn't have to use an instanceof here, although I see it is somewhat difficult, so does not seem necessary.

Reply:
right there are not many options